### PR TITLE
Api/request system in parts

### PR DESCRIPTION
**Pull request by marvin-bitterlich**
independent stuff

Review please

### CODE REVIEW

**Comment by MinecoloniesSonar** [comment]
SonarQube analysis reported 25 issues
* ![MINOR][MINOR] 25 minor

#### Top 10 extra issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![MINOR][MINOR] [blocks](https://github.com/Minecolonies/minecolonies/blob/57e3886dd30146bdadc94555e46ca48b906b8da2/src/main/java/com/minecolonies/coremod/blocks): Add a 'package-info.java' file to document the 'blocks' package [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1228)
1. ![MINOR][MINOR] [gui](https://github.com/Minecolonies/minecolonies/blob/57e3886dd30146bdadc94555e46ca48b906b8da2/src/main/java/com/minecolonies/coremod/client/gui): Add a 'package-info.java' file to document the 'gui' package [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1228)
1. ![MINOR][MINOR] [render](https://github.com/Minecolonies/minecolonies/blob/57e3886dd30146bdadc94555e46ca48b906b8da2/src/main/java/com/minecolonies/coremod/client/render): Add a 'package-info.java' file to document the 'render' package [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1228)
1. ![MINOR][MINOR] [colony](https://github.com/Minecolonies/minecolonies/blob/57e3886dd30146bdadc94555e46ca48b906b8da2/src/main/java/com/minecolonies/coremod/colony): Add a 'package-info.java' file to document the 'colony' package [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1228)
1. ![MINOR][MINOR] [buildings](https://github.com/Minecolonies/minecolonies/blob/57e3886dd30146bdadc94555e46ca48b906b8da2/src/main/java/com/minecolonies/coremod/colony/buildings): Add a 'package-info.java' file to document the 'buildings' package [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1228)
1. ![MINOR][MINOR] [permissions](https://github.com/Minecolonies/minecolonies/blob/57e3886dd30146bdadc94555e46ca48b906b8da2/src/main/java/com/minecolonies/coremod/colony/permissions): Add a 'package-info.java' file to document the 'permissions' package [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1228)
1. ![MINOR][MINOR] [creativetab](https://github.com/Minecolonies/minecolonies/blob/57e3886dd30146bdadc94555e46ca48b906b8da2/src/main/java/com/minecolonies/coremod/creativetab): Add a 'package-info.java' file to document the 'creativetab' package [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1228)
1. ![MINOR][MINOR] [entity](https://github.com/Minecolonies/minecolonies/blob/57e3886dd30146bdadc94555e46ca48b906b8da2/src/main/java/com/minecolonies/coremod/entity): Add a 'package-info.java' file to document the 'entity' package [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1228)
1. ![MINOR][MINOR] [basic](https://github.com/Minecolonies/minecolonies/blob/57e3886dd30146bdadc94555e46ca48b906b8da2/src/main/java/com/minecolonies/coremod/entity/ai/basic): Add a 'package-info.java' file to document the 'basic' package [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1228)
1. ![MINOR][MINOR] [farmer](https://github.com/Minecolonies/minecolonies/blob/57e3886dd30146bdadc94555e46ca48b906b8da2/src/main/java/com/minecolonies/coremod/entity/ai/citizen/farmer): Add a 'package-info.java' file to document the 'farmer' package [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1228)

[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

File: src/api/java/com/minecolonies/api/colony/requestsystem/StandardFactoryController.java [hunk 1]
```diff
@@ -0,0 +1,182 @@
...
+import net.minecraft.nbt.NBTTagCompound;
+import org.jetbrains.annotations.NotNull;
+
+import java.util.HashMap;
+
+/**
+ * Default implementation of a FactoryController
+ * Singleton.
+ */
+public class StandardFactoryController implements IFactoryController
```
Comment (marchermans) [inline src/api/java/com/minecolonies/api/colony/requestsystem/StandardFactoryController.java]: This is not in Union with the other PR's were this is already final

File: src/api/java/com/minecolonies/api/colony/requestsystem/factory/IFactoryController.java [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+     */
+    default <Input> IFactory<Input, ?> getFactoryForInput(@NotNull String className) throws IllegalArgumentException
+    {
+        //Simple default implementation grabs the class from a given name and casts it to the proper type.
+        //Any exceptions thrown before actual request is made gets wrapped.
+        try
+        {
+            return getFactoryForInput((Class<? extends Input>) Class.forName(className));
+        }
+        catch (IllegalArgumentException ex)
```
Comment (marchermans) [inline src/api/java/com/minecolonies/api/colony/requestsystem/factory/IFactoryController.java]: Nope i actually forcefully want to rethrow here. I singled out the IllegalArgumentExcetption as it is part of the method declaration.

File: src/api/java/com/minecolonies/api/colony/requestsystem/requestable/Weapon.java [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+        this.result = result;
+    }
+
+    /**
+     * The weapon type of the Weapon requested
+     *
+     * @return The weapon type of the weapon requested.
+     */
+    @NotNull
+    private WeaponType getType()
```
Comment (marchermans) [inline src/api/java/com/minecolonies/api/colony/requestsystem/requestable/Weapon.java]: Oeps IDEA autoconstructor seems to have made this private. This should be public.

**Review by marchermans — CHANGES_REQUESTED** [review]
Some Things Need to be merged with the other prs for the final Stuff.

I overruled one of codacities comments.

**Comment by MinecoloniesSonar** [comment]
SonarQube analysis reported 25 issues
* ![MINOR][MINOR] 25 minor

#### Top 10 extra issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![MINOR][MINOR] [blocks](https://github.com/Minecolonies/minecolonies/blob/8cf267315bd69376e569f1d30dc335fc804e4ce8/src/main/java/com/minecolonies/coremod/blocks): Add a 'package-info.java' file to document the 'blocks' package [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1228)
1. ![MINOR][MINOR] [gui](https://github.com/Minecolonies/minecolonies/blob/8cf267315bd69376e569f1d30dc335fc804e4ce8/src/main/java/com/minecolonies/coremod/client/gui): Add a 'package-info.java' file to document the 'gui' package [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1228)
1. ![MINOR][MINOR] [render](https://github.com/Minecolonies/minecolonies/blob/8cf267315bd69376e569f1d30dc335fc804e4ce8/src/main/java/com/minecolonies/coremod/client/render): Add a 'package-info.java' file to document the 'render' package [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1228)
1. ![MINOR][MINOR] [colony](https://github.com/Minecolonies/minecolonies/blob/8cf267315bd69376e569f1d30dc335fc804e4ce8/src/main/java/com/minecolonies/coremod/colony): Add a 'package-info.java' file to document the 'colony' package [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1228)
1. ![MINOR][MINOR] [buildings](https://github.com/Minecolonies/minecolonies/blob/8cf267315bd69376e569f1d30dc335fc804e4ce8/src/main/java/com/minecolonies/coremod/colony/buildings): Add a 'package-info.java' file to document the 'buildings' package [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1228)
1. ![MINOR][MINOR] [permissions](https://github.com/Minecolonies/minecolonies/blob/8cf267315bd69376e569f1d30dc335fc804e4ce8/src/main/java/com/minecolonies/coremod/colony/permissions): Add a 'package-info.java' file to document the 'permissions' package [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1228)
1. ![MINOR][MINOR] [creativetab](https://github.com/Minecolonies/minecolonies/blob/8cf267315bd69376e569f1d30dc335fc804e4ce8/src/main/java/com/minecolonies/coremod/creativetab): Add a 'package-info.java' file to document the 'creativetab' package [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1228)
1. ![MINOR][MINOR] [entity](https://github.com/Minecolonies/minecolonies/blob/8cf267315bd69376e569f1d30dc335fc804e4ce8/src/main/java/com/minecolonies/coremod/entity): Add a 'package-info.java' file to document the 'entity' package [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1228)
1. ![MINOR][MINOR] [basic](https://github.com/Minecolonies/minecolonies/blob/8cf267315bd69376e569f1d30dc335fc804e4ce8/src/main/java/com/minecolonies/coremod/entity/ai/basic): Add a 'package-info.java' file to document the 'basic' package [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1228)
1. ![MINOR][MINOR] [farmer](https://github.com/Minecolonies/minecolonies/blob/8cf267315bd69376e569f1d30dc335fc804e4ce8/src/main/java/com/minecolonies/coremod/entity/ai/citizen/farmer): Add a 'package-info.java' file to document the 'farmer' package [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1228)

[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

**Pull request closed without merging**